random-beans-validation/src/main/java/io/github/benas/randombeans/validation/PastAnnotationHandler.java

```diff
@@ -23,27 +23,28 @@
  */
 package io.github.benas.randombeans.validation;
 
+import io.github.benas.randombeans.EnhancedRandomBuilder;
+import io.github.benas.randombeans.api.EnhancedRandom;
 import io.github.benas.randombeans.api.Randomizer;
-import io.github.benas.randombeans.randomizers.range.DateRangeRandomizer;
 import io.github.benas.randombeans.util.Constants;
 
 import java.lang.reflect.Field;
-import java.util.Calendar;
-import java.util.Date;
-import java.util.Random;
+import java.time.LocalDate;
 
 class PastAnnotationHandler implements BeanValidationAnnotationHandler {
 
-    private final Random random;
+    private final long seed;
 
     public PastAnnotationHandler(long seed) {
-        random = new Random(seed);
+        this.seed = seed;
     }
 
     @Override
     public Randomizer<?> getRandomizer(Field field) {
-        Calendar calendar = Calendar.getInstance();
-        calendar.add(Calendar.YEAR, -Constants.DEFAULT_DATE_RANGE);
-        return new DateRangeRandomizer(calendar.getTime(), new Date(), random.nextLong());
+        EnhancedRandom enhancedRandom = EnhancedRandomBuilder.aNewEnhancedRandomBuilder()
```

Comment:
I think this delegate `enhancedRandom` can be a field an initialized in the constructor. This avoids creating it each time we call `getRandomizer`. Does this make sense?

Reply:
That was my first idea, but moving the code to the constructor results in:

```
java.lang.StackOverflowError
	at java.lang.String.endsWith(String.java:1449)
	at java.io.FilePermission$1.run(FilePermission.java:207)
	at java.io.FilePermission$1.run(FilePermission.java:203)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.io.FilePermission.init(FilePermission.java:203)
	at java.io.FilePermission.<init>(FilePermission.java:277)
	at sun.net.www.protocol.file.FileURLConnection.getPermission(FileURLConnection.java:228)
	at sun.net.www.protocol.jar.JarFileFactory.getPermission(JarFileFactory.java:166)
	at sun.net.www.protocol.jar.JarFileFactory.getCachedJarFile(JarFileFactory.java:136)
	at sun.net.www.protocol.jar.JarFileFactory.get(JarFileFactory.java:91)
	at sun.net.www.protocol.jar.JarURLConnection.connect(JarURLConnection.java:122)
	at sun.net.www.protocol.jar.JarURLConnection.getInputStream(JarURLConnection.java:152)
	at java.net.URL.openStream(URL.java:1045)
	at java.util.ServiceLoader.parse(ServiceLoader.java:304)
	at java.util.ServiceLoader.access$200(ServiceLoader.java:185)
	at java.util.ServiceLoader$LazyIterator.hasNextService(ServiceLoader.java:357)
	at java.util.ServiceLoader$LazyIterator.hasNext(ServiceLoader.java:393)
	at java.util.ServiceLoader$1.hasNext(ServiceLoader.java:474)
	at java.lang.Iterable.forEach(Iterable.java:74)
	at io.github.benas.randombeans.EnhancedRandomBuilder.loadRegistries(EnhancedRandomBuilder.java:454)
	at io.github.benas.randombeans.EnhancedRandomBuilder.setupRandomizerRegistries(EnhancedRandomBuilder.java:447)
	at io.github.benas.randombeans.EnhancedRandomBuilder.build(EnhancedRandomBuilder.java:432)
	at io.github.benas.randombeans.validation.PastAnnotationHandler.<init>(PastAnnotationHandler.java:44)
	at io.github.benas.randombeans.validation.BeanValidationRandomizerRegistry.init(BeanValidationRandomizerRegistry.java:57)
	at io.github.benas.randombeans.EnhancedRandomBuilder.lambda$setupRandomizerRegistries$0(EnhancedRandomBuilder.java:448)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at io.github.benas.randombeans.EnhancedRandomBuilder.setupRandomizerRegistries(EnhancedRandomBuilder.java:448)
	at io.github.benas.randombeans.EnhancedRandomBuilder.build(EnhancedRandomBuilder.java:432)
	at io.github.benas.randombeans.validation.PastAnnotationHandler.<init>(PastAnnotationHandler.java:44)
	at io.github.benas.randombeans.validation.BeanValidationRandomizerRegistry.init(BeanValidationRandomizerRegistry.java:57)
	at io.github.benas.randombeans.EnhancedRandomBuilder.lambda$setupRandomizerRegistries$0(EnhancedRandomBuilder.java:448)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at io.github.benas.randombeans.EnhancedRandomBuilder.setupRandomizerRegistries(EnhancedRandomBuilder.java:448)
	at io.github.benas.randombeans.EnhancedRandomBuilder.build(EnhancedRandomBuilder.java:432)
	at io.github.benas.randombeans.validation.PastAnnotationHandler.<init>(PastAnnotationHandler.java:44)
	at io.github.benas.randombeans.validation.BeanValidationRandomizerRegistry.init(BeanValidationRandomizerRegistry.java:57)
	at io.github.benas.randombeans.EnhancedRandomBuilder.lambda$setupRandomizerRegistries$0(EnhancedRandomBuilder.java:448)
	...
```

Reply:
ok I see. Can we at least create it only once? Something like:

```java
class PastAnnotationHandler implements BeanValidationAnnotationHandler {

    private final long seed;
    private EnhancedRandom enhancedRandom;

    public PastAnnotationHandler(long seed) {
        this.seed = seed;
    }

    @Override
    public Randomizer<?> getRandomizer(Field field) {
        if(enhancedRandom == null) {
            LocalDate now = LocalDate.now();
            enhancedRandom = EnhancedRandomBuilder.aNewEnhancedRandomBuilder()
                    .seed(seed)
                    .dateRange(now.minusYears(Constants.DEFAULT_DATE_RANGE), now)
                    .build();
        }
        return ((Randomizer<Object>) () -> enhancedRandom.nextObject(field.getType()));
    }
}
```

Reply:
Sure, but then it is no longer thread-safe. If that is o.k. I have no objections.

Reply:
That's true for all current randomizers if I'm not mistaken. So it should be ok.

Reply:
`java.util.Random` is thread-safe. Therefor most (if not all) Randomizers are thread-safe after construction. I do not know if this matters.

Reply:
I have update the pull request to include lazy initialization of the `EnhancedRandom` field.